core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -77,6 +78,7 @@ class DelayedFetch(delayMs: Long,
    * Case E: This broker is the leader, but the requested epoch is now fenced
    * Case F: The fetch offset locates not on the last segment of the log
    * Case G: The accumulated bytes from all the fetching partitions exceeds the minimum bytes
+   * Case H: A diverging epoch was found, return response to trigger truncation
```

Comment:
Good catch here and in `FetchSession`. Do you think we should consider doing these fixes separately so that we can get them into 2.7? Otherwise it might be difficult to tie this behavior to the 2.7 IBP.

Reply:
Makes sense, will submit another PR with just those changes.

Reply:
PR with the changes in DelayedFetch and FetchSession: https://github.com/apache/kafka/pull/9434

Reply:
PR with the changes in DelayedFetch and FetchSession: https://github.com/apache/kafka/pull/9434